quinn-proto/src/connection/streams/state.rs

```diff
@@ -797,24 +818,40 @@ impl StreamsState {
         expanded
     }
 
+    pub(super) fn max_send_data(&self, id: &StreamId) -> VarInt {
+        let remote = self.side != id.initiator();
+        match id.dir() {
+            Dir::Uni => self.initial_max_stream_data_uni,
+            // Remote/local appear reversed here because the transport parameters are named from
+            // the perspective of the peer.
+            Dir::Bi if remote => self.initial_max_stream_data_bidi_local,
+            Dir::Bi => self.initial_max_stream_data_bidi_remote,
+        }
+    }
+
     pub(super) fn insert(&mut self, remote: bool, id: StreamId) {
         let bi = id.dir() == Dir::Bi;
+        // bidirectional OR (unidirectional AND NOT remote)
         if bi || !remote {
-            let max_data = match id.dir() {
-                Dir::Uni => self.initial_max_stream_data_uni,
-                // Remote/local appear reversed here because the transport parameters are named from
-                // the perspective of the peer.
-                Dir::Bi if remote => self.initial_max_stream_data_bidi_local,
-                Dir::Bi => self.initial_max_stream_data_bidi_remote,
-            };
-            let stream = Send::new(max_data);
-            assert!(self.send.insert(id, stream).is_none());
+            if remote {
+                assert!(self.send.insert(id, None).is_none());
+            } else {
+                assert!(self
```

Comment:
Could we get away without eagerly allocating locally-initiated streams? Downstream code might want to e.g. reserve certain stream IDs without using them immediately.